### PR TITLE
Feature/jc 659 anime list management

### PR DESCRIPTION
**Pull request by sherstkov**
Added anime list management features like: type filters, `title_eng`, `aired.start`, `status` sorting and pagination.

### CODE REVIEW

File: apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html [hunk 1]
```diff
@@ -1 +1,44 @@
...
+		class="pagination"
+		[length]="animeTotal"
+		[pageSize]="pageSize$ | async"
+		[showFirstLastButtons]="true"
+		[pageSizeOptions]="[5, 10, 15]"
+		[pageIndex]="currentPage$ | async"
+		(page)="handlePageEvent($event)"
+		[disabled]="isLoading$ | async"
+	>
+	</mat-paginator>
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html:43]:
```suggestion
	<camp-anime-table class="anime-table" [animeList]="animeList$ | async" />
	
	<mat-paginator
		class="pagination"
		[length]="animeTotal"
		[pageSize]="pageSize$ | async"
		[showFirstLastButtons]="true"
		[pageSizeOptions]="[5, 10, 15]"
		[pageIndex]="currentPage$ | async"
		(page)="handlePageEvent($event)"
		[disabled]="isLoading$ | async"
	/>
```

File: apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html [hunk 2]
```diff
@@ -1 +1,44 @@
...
+						{{ type }}
+					</mat-option>
+				</mat-select>
+			</mat-form-field>
+			<form class="search-form" (ngSubmit)="onSearch()">
+				<mat-form-field appearance="outline" subscriptSizing="dynamic">
+					<input
+						matInput
+						type="text"
+						[(ngModel)]="searchValue"
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html:23]:
```suggestion
						[(ngModel)]="searchValue"
```

Let's use `formControls`

File: apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html [hunk 3]
```diff
@@ -1 +1,44 @@
...
+					(selectionChange)="onFilter($event)"
+					placeholder="Type of an anime"
+					multiple
+				>
+					<mat-option *ngFor="let type of animeTypes" [value]="type">
+						{{ type }}
+					</mat-option>
+				</mat-select>
+			</mat-form-field>
+			<form class="search-form" (ngSubmit)="onSearch()">
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html:18]: https://htmlacademy.ru/blog/html/search

Not long ago a new tag appeared

File: apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.ts [hunk 4]
```diff
@@ -11,12 +24,112 @@ import { AnimeTableComponent } from './components/anime-table/anime-table.compon
...
+	protected isLoading$ = this.animeService.observableIsLoading$;
+
+	/** Anime count. */
+	protected animeTotal = AnimeParams.defaultValues.animeTotal;
+
+	/** Search input value. */
+	protected searchValue = AnimeParams.defaultValues.searchValue;
+
+	/** Anime type values. */
+	protected animeTypes = Object.values(AnimeType);
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.ts:69]:
```suggestion
	protected readonly isLoading$ = this.animeService.observableIsLoading$;

	/** Anime count. */
	protected readonly animeTotal = AnimeParams.defaultValues.animeTotal;

	/** Search input value. */
	protected readonly searchValue = AnimeParams.defaultValues.searchValue;

	/** Anime type values. */
	protected readonly animeTypes = Object.values(AnimeType);
```

We follow the immutability approach; most often, class fields are immutable. Let's make decisions that will be based on this principle

File: apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html (hunk 3 above)
Comment (sherstkov) [reply apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html:18]: Angular didn't know yet :(
```
1. If 'search' is an Angular component, then verify that it is included in the '@Component.imports' of this component.
```

I can use `CUSTOM_ELEMENTS_SCHEMA` in the component, but should I?

Comment (andrey-gazzaev) [reply apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html]: lol

File: apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.ts (hunk 4 above)
Comment (sherstkov) [reply apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.ts]: Hope i get it right :crossed_fingers:

File: apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts [hunk 5]
```diff
@@ -12,11 +14,14 @@ import { Anime } from '@js-camp/core/models/anime';
...
 	styleUrl: './anime-table.component.css',
 	changeDetection: ChangeDetectionStrategy.OnPush,
-	imports: [CommonModule, MatTableModule, EmptyPipe, BasicProgressSpinnerComponent],
+	imports: [CommonModule, MatTableModule, EmptyPipe, MatSortModule, BasicProgressSpinnerComponent],
 })
 export class AnimeTableComponent {
+
 	/** Anime list. */
-	@Input() public animeList: Anime[] | null = null;
+	@Input() public animeList: readonly Anime[] | null = [];
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts:22]:
```suggestion
	@Input() 
	public animeList: readonly Anime[] = [];
```

Why might the value be `null`? Why does this component even care that the value might be different from the entity the component is working with?

Leave the data handling to the consumers of this component.

File: apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts [hunk 6]
```diff
@@ -12,11 +14,14 @@ import { Anime } from '@js-camp/core/models/anime';
...
+
+	private readonly animeService: AnimeService = inject(AnimeService);
 
 	/** Columns names. */
 	protected readonly displayedColumns = ['image', 'englishTitle', 'japaneseTitle', 'airedStart', 'type', 'status'] as const;
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts:27]: Let's use the enum

File: apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts [hunk 7]
```diff
@@ -26,9 +31,17 @@ export class AnimeTableComponent {
...
 	 */
 	protected trackByAnimeId(_index: number, item: Anime): Anime['id'] {
 		return item.id;
 	}
+
+	/**
+	 * Handles the change in sorting.
+	 *
+	 * @param event - The event object containing the sorting information.
+	 */
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts:43]:
```suggestion
	/**
	 * Handles the sort change.
	 * @param sort - New sort value.
	 */
	 protected handleSortChange(sort: Sort): void
```

File: apps/angular/src/core/services/anime.service.ts [hunk 8]
```diff
@@ -1,30 +1,116 @@
...
+export class AnimeService implements OnDestroy {
 	private readonly http = inject(HttpClient);
 
-	private readonly urlService = inject(ApiUrlService);
+	private readonly apiUrlService = inject(ApiUrlService);
+
+	private readonly queryParamsService = inject(QueryParamsService);
+
+	/** Subject for managing unsubscribing. */
+	private readonly destroy$ = new Subject<void>();
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/anime.service.ts:31]: let's use [destroyRef ](https://angular.dev/api/core/DestroyRef) with [takeUntilDestroyed](https://angular.dev/api/core/rxjs-interop/takeUntilDestroyed#)

File: apps/angular/src/core/services/anime.service.ts [hunk 9]
```diff
@@ -1,30 +1,116 @@
...
+
+	/** Anime parameters subject. */
+	private readonly animeParams$: BehaviorSubject<AnimeParams> = new BehaviorSubject(new AnimeParams());
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/anime.service.ts:34]:
```suggestion
	private readonly _animeParams$ = new BehaviorSubject<AnimeParams>(new AnimeParams());
```

File: apps/angular/src/core/services/anime.service.ts [hunk 10]
```diff
@@ -1,30 +1,116 @@
...
+
+	/** Anime parameters observable for monitoring them outside the service. */
+	public readonly observableAnimeParams$ = this.animeParams$.asObservable();
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/anime.service.ts:37]:
```suggestion
	public readonly animeParams$ = this.animeParams$.asObservable();
```

File: apps/angular/src/core/services/anime.service.ts [hunk 11]
```diff
@@ -1,30 +1,116 @@
...
+
+	/** Loading state subject. */
+	private readonly isLoading$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/anime.service.ts:40]:
```suggestion
	private readonly isLoading$ = new BehaviorSubject(false);
```

File: apps/angular/src/core/services/anime.service.ts [hunk 12]
```diff
@@ -1,30 +1,116 @@
...
+
+	/** Loading state observable for monitoring it outside the service. */
+	public readonly observableIsLoading$ = this.isLoading$.asObservable();
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/anime.service.ts:43]: same

File: apps/angular/src/core/services/anime.service.ts [hunk 13]
```diff
@@ -1,30 +1,116 @@
...
+				return this.http.get<PaginationDto<AnimeDto>>(this.apiUrlService.animeListPath, { params: httpParams }).pipe(
+					map(pagination => PaginationMapper.fromDto(
+						pagination,
+						animeDto => AnimeMapper.fromDto(animeDto),
+					)),
+					tap(() => this.isLoading$.next(false)),
+				);
+			}),
+			takeUntil(this.destroy$),
+		);
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/anime.service.ts:61]: You have an abstraction leak. In your service, there is work with URL query, with http parameters and with models and DTOs.

These are all entities of different levels:
URL query is the UI level
Model is the domain level
http params and Dto are the response level

File: apps/angular/src/core/services/anime.service.ts [hunk 14]
```diff
@@ -1,30 +1,116 @@
...
+import { QueryParamsService } from './query-params.service';
 
 /**
  * Anime service to interact with the API.
  */
 @Injectable({
 	providedIn: 'root',
 })
-export class AnimeService {
+export class AnimeService implements OnDestroy {
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/anime.service.ts:23]:
```suggestion
export class AnimeService {
```
The service has no life hooks

File: apps/angular/src/core/services/query-params.service.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+			queryParams: params,
+			queryParamsHandling: 'merge',
+		});
+	}
+
+	/**
+	 * Get an instance of HTTPParams from params.
+	 * @param params Page parameters.
+	 *  */
+	public getHttpParams(params: AnimeParams): HttpParams {
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/query-params.service.ts:30]:
```suggestion
	public getQueryParams(params: AnimeParams): HttpParams {
```
Why do you use response level terms at the UI level?

File: apps/angular/src/core/services/query-params.service.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+		const animeParams = { ...AnimeParamsMapper.toDto(params) };
+		return new HttpParams({ fromObject: animeParams });
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/query-params.service.ts:32]: Judging by the name, the service can work with any data structures that need to be written in the URL, but in fact this service only works with AnimeParams

File: libs/core/mappers/based-params.mapper.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+import { AnimeTypeMapper } from './anime-type.mapper';
+
+export namespace AnimeParamsMapper {
+
+	/**
+	 * Converts a string array to AnimeType array.
+	 *
+	 * @param array The string array to convert.
+	 */
+	export function toArrayAnimeType(array: string[]): AnimeType[] {
```
Comment (andrey-gazzaev) [inline libs/core/mappers/based-params.mapper.ts:14]:
```suggestion
	export function toAnimeType(array: readonly string[]): AnimeType[] {
```

File: libs/core/mappers/based-params.mapper.ts [hunk 18]
```diff
@@ -0,0 +1,39 @@
+import { AnimeParamsDto } from '../dtos/based-params.dto';
```
Comment (andrey-gazzaev) [inline libs/core/mappers/based-params.mapper.ts:1]: Rename the file

File: libs/core/mappers/based-params.mapper.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+		return isValid ? array as AnimeType[] : AnimeParams.defaultValues.filterByType;
+	}
+
+	/**
+	 * Maps model to dto.
+	 * @param model AnimeParams model.
+	 */
+	export function toDto(model: AnimeParams): AnimeParamsDto {
+		const typeFilter = Array.isArray(model.filterByType) ?
+			model.filterByType.map(AnimeTypeMapper.toDto).join(',') :
```
Comment (andrey-gazzaev) [inline libs/core/mappers/based-params.mapper.ts:26]:
```suggestion
			model.filterByType.map((type) => AnimeTypeMapper.toDto(type)).join(',') :
```
Be careful, in this case you have lost the context

File: libs/core/models/based-params.ts [hunk 20]
```diff
@@ -0,0 +1,42 @@
+import { AnimeType } from './anime-type';
+import { Immerable, OmitImmerable } from './immerable';
+
+/** Anime params. */
+export class AnimeParams extends Immerable {
+
+	/** Default values for AnimeParams. */
+	public static readonly defaultValues = {
```
Comment (andrey-gazzaev) [inline libs/core/models/based-params.ts:8]:
```suggestion
	public static readonly DEFAULT_VALUE = {
```

File: libs/core/models/based-params.ts [hunk 21]
```diff
@@ -0,0 +1,42 @@
+import { AnimeType } from './anime-type';
+import { Immerable, OmitImmerable } from './immerable';
+
+/** Anime params. */
+export class AnimeParams extends Immerable {
```
Comment (andrey-gazzaev) [inline libs/core/models/based-params.ts:5]:
```suggestion
export class AnimeQuery extends Immerable {
```
It's similar to query

File: libs/core/models/based-params.ts [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+	public readonly pageSize: number;
+
+	/** Search input value. */
+	public readonly searchValue: string;
+
+	/** Sort order. */
+	public readonly sortOrder: string;
+
+	/** Filter values. */
+	public readonly filterByType: AnimeType[];
```
Comment (andrey-gazzaev) [inline libs/core/models/based-params.ts:30]:
```suggestion
	public readonly filterByTypes: readonly AnimeType[];
```

File: libs/core/models/based-params.ts [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+	public constructor(data: Partial<AnimeParamsData> = {}) {
```
Comment (andrey-gazzaev) [inline libs/core/models/based-params.ts:32]:
```suggestion
	public constructor(data: Partial<AnimeParamsData> = AnimeParams.DEFAULT_VALUE ) {
	super();
		this.pageIndex = data.pageIndex;
		this.pageSize = data.pageSize;
		this.searchValue = data.searchValue;
		this.sortOrder = data.sortOrder;
		this.filterByType = data.filterByType;
```

File: libs/theme/src/reset.css [hunk 24]
```diff
@@ -0,0 +1,9 @@
+*,
+*::before,
+*::after {
+  box-sizing: border-box;
+}
+
+* {
+  margin: 0;
+}
```
Comment (andrey-gazzaev) [inline libs/theme/src/reset.css:9]: You can use https://meyerweb.com/eric/tools/css/reset/

**Review by andrey-gazzaev — CHANGES_REQUESTED** [review]

File: apps/angular/src/core/services/anime.service.ts (hunk 8 above)
Comment (sherstkov) [reply apps/angular/src/core/services/anime.service.ts:31]: > let's use [takeUntilDestroyed](https://angular.dev/api/core/rxjs-interop/takeUntilDestroyed#)

Yes, but `takeUntilDestroyed` in the developer preview. Is this OK to use now?

File: apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts [hunk 25]
```diff
@@ -12,23 +16,63 @@ import { Anime } from '@js-camp/core/models/anime';
...
+export class AnimeTableComponent implements OnInit {
 	/** Anime list. */
-	@Input() public animeList: Anime[] | null = null;
+	@Input()
+	public animeList: readonly Anime[] = [];
+
+	private readonly pageParamsService = inject(PageParamsService);
+
+	/** Anime enum. */
+	protected readonly animeEnum = AnimeEnum;
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts:29]: What is enum in the context of a table? Maybe these are table columns?

File: apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts [hunk 26]
```diff
@@ -12,23 +16,63 @@ import { Anime } from '@js-camp/core/models/anime';
...
+	/** Sorting order subject. */
+	protected readonly sortOrder$ = new BehaviorSubject({ active: '', direction: '' } as SortOrder);
+
+	/** @inheritdoc */
+	public ngOnInit(): void {
+		this.animeParams$.pipe(
+			take(1),
+		).subscribe(pageParams => {
+			this.sortOrder$.next(this.parseSortOrder(pageParams.sortOrder));
+		});
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts:46]: Why does your table process sorting or filter values? Shouldn't the table just show the content?

File: apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts [hunk 27]
```diff
@@ -12,23 +16,63 @@ import { Anime } from '@js-camp/core/models/anime';
...
 		return item.id;
 	}
+
+	/**
+	 * Handles the sort change.
+	 *
+	 * @param sort New sort value.
+	 */
+	protected handleSortChange(sort: Sort): void {
+		this.pageParamsService.changeSortParams(sort);
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts:65]: same

Comment (andrey-gazzaev) [reply apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts:65]: The fact is that this architecture is not obvious. You use services that work with filters from different pieces of code. Thus, you disrupt the flow of data from component to component. 

Now you can't guarantee that the service data will not be changed by any other component with this approach.

File: apps/angular/src/core/services/anime.service.ts [hunk 28]
```diff
@@ -17,14 +22,29 @@ import { ApiUrlService } from './api-url.service';
...
+			switchMap(params => {
+				const httpParams = AnimeParamsMapper.toAnimeHttp(params);
+
+				return this.http.get<PaginationDto<AnimeDto>>(this.apiUrlService.animeListPath, { params: httpParams }).pipe(
+					map(pagination => PaginationMapper.fromDto(
+						pagination,
+						animeDto => AnimeMapper.fromDto(animeDto),
+					)),
+				);
+			}),
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/anime.service.ts:46]: same

Comment (andrey-gazzaev) [reply apps/angular/src/core/services/anime.service.ts:46]: Let's make the data transfer from the component to the service component explicit.

File: apps/angular/src/core/services/page-params.service.ts [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+
+import { QueryParamsService } from './query-params.service';
+
+/**
+ * Page parameters service.
+ */
+@Injectable({
+	providedIn: 'root',
+})
+export class PageParamsService {
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/page-params.service.ts:16]: You are still using a too abstract name for the service, although it only works with anime parameters

File: apps/angular/src/core/services/page-params.service.ts [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+	public changeSortParams(sortValue: Sort): void {
+		let sortOrder = '';
+		if (sortValue.direction === 'asc') {
+			sortOrder = sortValue.active;
+		}
+		if (sortValue.direction === 'desc') {
+			sortOrder = `-${sortValue.active}`;
+		}
+		this.changeAnimeParams({ sortOrder });
+	}
```
Comment (andrey-gazzaev) [inline apps/angular/src/core/services/page-params.service.ts:78]: Why do you need it?

File: libs/core/models/sort-order.ts [hunk 31]
```diff
@@ -0,0 +1,11 @@
...
+
+/** Sorting order object. */
+export type SortOrder = {
+
+	/** Sorted field. */
+	active: string;
+
+	/** Sorting direction. */
+	direction: SortDirection;
+};
```
Comment (andrey-gazzaev) [inline libs/core/models/sort-order.ts:11]: How is order different from direction?

File: apps/angular/src/core/services/anime.service.ts (hunk 28 above)
Comment (sherstkov) [reply apps/angular/src/core/services/anime.service.ts:46]: What do you think of this approach? Now we are more explicit that we are using params as an observable and not changing params.
```suggestion
private readonly animeParams$ = this.pageParamsService.getAnimeParams();
```

File: apps/angular/src/core/services/page-params.service.ts (hunk 30 above)
Comment (sherstkov) [reply apps/angular/src/core/services/page-params.service.ts:78]: This way it's clear to see how sorting changes. It's basically 3 states: empty sort, ascending sort and descending sort. I made the docs more explicit about this.

File: apps/angular/src/core/services/anime.service.ts (hunk 28 above)
Comment (andrey-gazzaev) [reply apps/angular/src/core/services/anime.service.ts]: You do not need to store filter parameters in the service at all. Just pass them as a method argument

File: apps/angular/src/core/services/page-params.service.ts (hunk 30 above)
Comment (andrey-gazzaev) [reply apps/angular/src/core/services/page-params.service.ts]: Why can't you just use a flat object?
```{ page, pageSize, search, types, field, direction }```

File: libs/core/mappers/anime-params.mapper.ts [hunk 32]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	/**
+	 * Maps anime query params to http params.
+	 * @param params Query parameters.
+	 */
+	export function toAnimeHttp(params: AnimeQuery): HttpParams {
+		const animeParams = { ...toDto(params) };
+		return new HttpParams({ fromObject: animeParams });
+	}
```
Comment (andrey-gazzaev) [inline libs/core/mappers/anime-params.mapper.ts:47]: Can this method work with other structures? Except Anime Query. If yes, then it looks like a utility function

File: libs/core/models/sort-order.ts [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+
+	/**
+	 * Sorting field by which data is to be sorted (like 'name' or 'date').
+	 */
+	active: string;
+
+	/**
+	 * Sorting direction (ascending, descending or no order).
+	 */
+	direction: SortDirection;
```
Comment (andrey-gazzaev) [inline libs/core/models/sort-order.ts:14]:
```suggestion
	/** Sorting field by which data is to be sorted (like 'name' or 'date'). */
	readonly active: string;

	/** Sorting direction (ascending, descending or no order).  */
	readonly direction: SortDirection;
```

**Review by andrey-gazzaev — COMMENTED** [review]
Before we go any further, let's fix the problem in Anime Service

File: apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html [hunk 34]
```diff
@@ -1 +1,48 @@
...
+			[pageSize]="pageSize$ | async"
+			[showFirstLastButtons]="true"
+			[pageSizeOptions]="[5, 10, 15]"
+			[pageIndex]="currentPage$ | async"
+			(page)="handlePageEvent($event)"
+			[disabled]="animeList.results.length === 0"
+			/>
+		</div>
+	} @else {
+		<camp-basic-progress-spinner></camp-basic-progress-spinner>
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/anime-dashboard.component.html:46]:
```suggestion
		<camp-basic-progress-spinner />
```

File: apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.html [hunk 35]
```diff
@@ -1,53 +1,46 @@
...
-				<th mat-header-cell *matHeaderCellDef> Status </th>
-				<td mat-cell *matCellDef="let element"> {{element.status}} </td>
-			</ng-container>
-
-			<tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
-			<tr mat-row *matRowDef="let row; columns: displayedColumns;"></tr>
-		</table>
-	}
+@if (animeList.length === 0){
+	<p class="anime-list__empty-message"> No anime to display. </p>
```
Comment (andrey-gazzaev) [inline apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.html:3]: https://material.angular.io/components/table/api#MatNoDataRow

**Review by andrey-gazzaev — APPROVED** [review]
Well done 🔥

File: apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts [hunk 36]
```diff
@@ -12,21 +14,35 @@ import { Anime } from '@js-camp/core/models/anime';
...
+	@Input()
+	public animeList: readonly Anime[] = [];
+
+	/** Initial sort params. */
+	@Input()
+	public sortParams: Sort = AnimeQuery.DEFAULT_VALUES.sort;
+
+	/** Emits new sort value. */
+	@Output()
+	public sortChange = new EventEmitter<Sort>();
```
Comment (karbik-creator) [inline apps/angular/src/app/features/anime-dashboard/components/anime-table/anime-table.component.ts:30]: Please add readonly
https://wiki.saritasa.rocks/frontend/frameworks/angular/best-practices#prefer-readonly-outputs

File: apps/angular/src/core/services/anime.service.ts [hunk 37]
```diff
@@ -17,14 +23,22 @@ import { ApiUrlService } from './api-url.service';
...
 	/**
 	 * Get a list of anime from the API.
+	 *
+	 * @param params Anime query parameters.
 	 */
-	public getAnimeList(): Observable<Anime[]> {
-		return this.http.get<PaginationDto<AnimeDto>>(this.urlService.animeListPath).pipe(
-			map(animeDto => animeDto.results.map(anime => AnimeMapper.fromDto(anime))),
+	public getAnimeList(params: AnimeQuery): Observable<Pagination<Anime>> {
+		const animeParams = { ...AnimeParamsMapper.toDto(params) };
```
Comment (karbik-creator) [inline apps/angular/src/core/services/anime.service.ts:34]:
```suggestion
		const animeParams = AnimeParamsMapper.toDto(params);
```

File: apps/angular/src/core/services/page-anime-params.service.ts [hunk 38]
```diff
@@ -0,0 +1,79 @@
...
+			this.queryParamsService.changeQueryParams(animeQueryParams);
+		});
+	}
+
+	/**
+	 * Changes Subject of a pagination parameters.
+	 * @param paginationParams Pagination parameters.
+	 */
+	public changePaginationParams(paginationParams: AnimeQuery): void {
+		this.changeAnimeParams({ pageIndex: paginationParams.pageIndex, pageSize: paginationParams.pageSize });
```
Comment (karbik-creator) [inline apps/angular/src/core/services/page-anime-params.service.ts:49]:
```suggestion
		this.changeAnimeParams(paginationParams);
```

**Review by karbik-creator — APPROVED** [review]

**Review by pavelzadorin-saritasa — APPROVED** [review]

**Pull request merged**